dfastmi/io/map_file.py

```diff
@@ -0,0 +1,347 @@
+        ---------
+            map_file : Path
+                The path to the map file.
+        """
+        self._map_file = map_file
+        self._mesh2d_name = None
+        self._face_dimension_name = None
+    
+    @property
+    def node_x_coordinates(self) -> np.ndarray:
+        """Get the x-coordinates of the nodes.
+
+        Returns
+        -------
+        numpy.ndarray
+            Array with shape (N,) where N is the number of nodes.
+        """
+        return self._get_node_coordinate_data(["projection_x_coordinate", "longitude"])
+    
+    @property
+    def node_y_coordinates(self) -> np.ndarray:
+        """Get the y-coordinates of the nodes.
+
+        Returns
+        -------
+        numpy.ndarray
+            Array with shape (N,) where N is the number of nodes.
+        """
+        return self._get_node_coordinate_data(["projection_y_coordinate", "latitude"])
+    
+    @property
+    def face_node_connectivity(self) -> ma.masked_array:
+        """Get the face-node connectivity from the 2d mesh.
+
+        Returns
+        -------
+        ma.masked_array
+            Array with shape (N,M) where N is the number of faces and M the maximum number of nodes per face.
+            If not all the faces have the same number of nodes, a boolean mask is provided with shape (N,M) 
+            where each True value indicates a fill value.
+        """
+        with nc.Dataset(self._map_file) as rootgrp:
+            mesh2d = rootgrp.variables[self.mesh2d_name]
+            var_name = mesh2d.getncattr("face_node_connectivity")
+            var = rootgrp.variables[var_name]
+            data = var[...] - MapFile._get_start_index(var)
+
+        return data
+
+    @staticmethod
+    def _get_start_index(var) -> int:
+        if "start_index" in var.ncattrs():
+            return var.getncattr("start_index")
+        return 0
+    
+    def read_face_variable(self, 
+                      varname: str, 
+                      time_index_from_last: Optional[int] = None,
+    ) -> np.ndarray:
+        """
+        Read the last time step of any quantity defined at faces from a D-Flow FM map-file.
+
+        Arguments
+        ---------
+        varname : str
+            Name of the netCDF variable to be read.
+        time_index_from_last : Optional[int]
+            Time step offset index from the last time step written.
+
+        Raises
+        ------
+        ValueError
+            If the data file doesn't include a 2D mesh.
+            If it cannot uniquely identify the variable to be read.
+
+        Returns
+        -------
+        numpy.ndarray
+            1D data of the requested variable. If the variable is time-dependent, 
+            the time_index_from_last is used.
+        """
+        # open file
+        with nc.Dataset(self._map_file) as rootgrp:
+            location = "face"
+
+            # find any other variable by standard_name or long_name
+            var = rootgrp.get_variables_by_attributes(
```

Comment:
these two piece of code in this method can be split up, e.g. "get_variable_by_name" and "get_variable_data"

Reply:
Done